File: python-api/verify_the_binary.py
import subprocess
import platform
import sys
import json


def check_command_output(output) -> dict:
    """
    Will return a dict if output correctly formatted. Otherwise, raise exception
    """
    output_delimiter = 'OUTPUT:'
    if len(output.split(output_delimiter)) != 2:
        raise Exception("Error. Binary improperly working. Inspect the source code of this script to see why.")
    output_dict = output.split('OUTPUT:')[-1]
    try:
        output_dict = json.loads(output_dict)
        return output_dict
    except json.decoder.JSONDecodeError:
        raise Exception("Error. Binary improperly working. Inspect the source code of this script to see why.")


def main():

    # Name of the program
    _ = sys.argv[0]
    # The path to the binary
    if len(sys.argv) != 2:
        print("USAGE: python3 verify.py path/to/binary")
        exit()

    path_to_binary = sys.argv[1]

    if platform.system() in ['Windows', 'Linux', 'Darwin']:
        """
        Input Test #1
        ```
        ./test-executables/main
        ```
        Expected Output
        ```
        OUTPUT: {"data": {"error_message": "Invalid execution.", "error_code": -2}, "response_code": 400}
        ```
        """
        result = subprocess.run(
            [
                f'./{path_to_binary}'
            ],
            capture_output=True
        )
        output = check_command_output(result.stdout.decode())
        if output['response_code'] != 400:
            raise Exception("`response_code` must be 400.")

        """
        Input Test #2
        ```
        ./test-executables/main ping '{}' '{}'
        ```
        Expected Output
        ```
        OUTPUT: {"data": {}, "response_code": 200}
        ```
        """
        result = subprocess.run(
            [
                f'./{path_to_binary}',
                'ping',
                "'{}'",
                "'{}'"
            ],
            capture_output=True
        )
        output = check_command_output(result.stdout.decode())
        if output['response_code'] != 200:
            raise Exception("`response_code` must be 400.")

        print("*** All Tests Passed *** \nAdd more tests here to verify your binary. "
              "We highly recommend testing your API as you develop it. "
              "If you don't, it becomes almost impossible to make changes in the future because you don't have a "
              "quick and easy way to ensure that those changes did not break anything.")
    else:
        print(f"Unrecognized platform, `{platform.system()}`.")
        exit()


if __name__ == '__main__':
    main()

File: node-js-api/verifyTheBinary.js
const { spawnSync } = require( 'child_process' );

// `npm i child_process`


const checkCommandOutput = (output) => {
    /*
    Will return a dict if output correctly formatted. Otherwise, raise exception
    */
    
    const outputDelimiter = 'OUTPUT:'
    const splitArr = output.split(outputDelimiter);

    if (splitArr.length < 2) {
        throw Error("Error. Binary improperly working. Inspect the source code of this script to see why.")
    }
    let outputDict = splitArr[splitArr.length - 1];
    
    try {
        outputDict = JSON.parse(outputDict)
        console.log(outputDict);
        return outputDict
    } catch (error) {
        console.log(error);
        throw Error("Error. Binary improperly working. Inspect the source code of this script to see why.")
    }
}


const main = () => {

    const platform = process.platform;
    let path_to_binary;

    if (platform == 'win32') {
        path_to_binary = 'binaries/main-win.exe';
    } else if (platform == 'darwin') {
        path_to_binary = 'binaries/main-macos';
    } else if (platform == 'linux') {
        path_to_binary = 'binaries/main-linux';
    } else {
        throw Error("Unrecognized platform: " + platform);
    }

    console.log("NOTE: This file tests the binary at path `" + path_to_binary + "` but we're uploading the binary at `my-api/main`.")
    console.log("Thus, if the tests pass here but not after uploading, make sure that *both* binaries are uploaded.")

    /*
    Input Test #1
    ```
    ./test-executables/main
    ```
    Expected Output
    ```
    OUTPUT: {"data": {"error_message": "Invalid execution.", "error_code": -2}, "response_code": 400}
    ```
    */
    let result = spawnSync('./' + path_to_binary).stdout;
    result = String.fromCharCode(...result);
    let output = checkCommandOutput(result)
    if (output['response_code'] != 400) {
        throw Error("`response_code` must be 400.")
    }

    /*
    Input Test #2
    ```
    ./test-executables/main ping '{}' '{}'
    ```
    Expected Output
    ```
    OUTPUT: {"data": {}, "response_code": 200}
    ```
    */
    result = spawnSync('./' + path_to_binary, ['ping', JSON.stringify({}), JSON.stringify({})]).stdout;
    result = String.fromCharCode(...result);
    output = checkCommandOutput(result)
    if (output['response_code'] != 200) {
        throw Error("`response_code` must be 200.")
    }

    result = spawnSync('./' + path_to_binary, ['generatePublicKey', JSON.stringify({}), JSON.stringify({"startingWith": "A"})]).stdout;
        result = String.fromCharCode(...result);
        output = checkCommandOutput(result)
        if (output['response_code'] != 200) {
            throw Error("`response_code` must be 200.")
        }

    console.log("*** All Tests Passed *** \nAdd more tests here to verify your binary. We highly recommend testing your API as you develop it. If you don't, it becomes almost impossible to make changes in the future because you don't have a quick and easy way to ensure that those changes did not break anything.")
}


main()
